### PR TITLE
Initial State `NonLinearAnalysis`

### LINKED ISSUE #422: Set an initial `State` to solve the `StructuralAnalysis`
- [x] LinearStaticAnalysis
- [ ] NonLinearStaticAnalysis

    
```julia
s = Structure(msh_file, s_materials, s_boundary_conditions, s_entities)


# First analysis
first_analysis = LinearStaticAnalysis(s; NSTEPS=NSTEPS)
first_solution = solve!(sa)

# Second analysis
second_analysis = NonLinearStaticAnalysis(s; NSTEPS=NSTEPS, initial_state=last(states(first_solution)))
second_solution = solve!(sa)
```

### PR DESCRIPTION
**Pull request by mvanzulli**
Closes #422 

Requires #425  and #427

### CODE REVIEW

File: src/StructuralAnalyses/NonLinearStaticAnalyses.jl [hunk 1]
```diff
@@ -39,18 +39,24 @@ struct NonLinearStaticAnalysis{S<:AbstractStructure,LFV<:AbstractVector{<:Real}}
...
+
+    NonLinearStaticAnalysis(s, initial_state, λᵥ, ScalarWrapper(initial_step))
 end
 
 "Constructor for non linear static analysis given a final time (or load factor) and the number of steps."
 function NonLinearStaticAnalysis(s::AbstractStructure, t₁::Real=1.0; NSTEPS=10, initial_step::Int=1)
     t₀ = t₁ / NSTEPS
     λᵥ = collect(LinRange(t₀, t₁, NSTEPS))
-    return NonLinearStaticAnalysis(s, λᵥ; initial_step=initial_step)
+    NonLinearStaticAnalysis(s, λᵥ; initial_step=initial_step)
```
Comment (mforets) [inline src/StructuralAnalyses/NonLinearStaticAnalyses.jl]:
```suggestion
    NonLinearStaticAnalysis(s, λᵥ; initial_step)
```

**Review by mforets — APPROVED** [review]

File: src/StructuralAnalyses/NonLinearStaticAnalyses.jl [hunk 2]
```diff
@@ -39,18 +39,24 @@ struct NonLinearStaticAnalysis{S<:AbstractStructure,LFV<:AbstractVector{<:Real}}
...
+"Constructor for a non linear analysis with load factors, optional initial step and initial state."
+function NonLinearStaticAnalysis(s::S, λᵥ::LFV;
+                                 initial_state::StaticState=StaticState(s),
+                                 initial_step::Int=1) where {S<:AbstractStructure,
+                                                             LFV<:AbstractVector{<:Real}}
+    !(1 ≤ initial_step ≤ length(λᵥ)) &&
+        throw(ArgumentError("initial_step must be in [1, $(length(λᵥ))] but is: $initial_step."))
+    NonLinearStaticAnalysis(s, initial_state, λᵥ, ScalarWrapper(initial_step))
+
+    NonLinearStaticAnalysis(s, initial_state, λᵥ, ScalarWrapper(initial_step))
```
Comment (mforets) [inline src/StructuralAnalyses/NonLinearStaticAnalyses.jl]: same line twice?

**Comment by mforets** [comment]
About `ScalarWrapper`, in the current Julia version it is possible to leverage `const` annotation in struct definitions:

```julia
julia> mutable struct Foo
           const A::Float64
           x::Int64
       end

julia> f = Foo(1.0, 2)
Foo(1.0, 2)

julia> f.x = 5
5

julia> f.A = 2.0
ERROR: setfield!: const field .A of type Foo cannot be changed
Stacktrace:
 [1] setproperty!(x::Foo, f::Symbol, v::Float64)
   @ Base ./Base.jl:38
 [2] top-level scope
   @ REPL[4]:1
 [3] top-level scope
   @ ~/.julia/packages/Infiltrator/LtFao/src/Infiltrator.jl:726

julia> f
Foo(1.0, 5)
```

**Comment by mforets** [comment]
So we may as well remove the `ScalarWrapper` in favor of using the previous pattern.

**Comment by mvanzulli** [comment]
> So we may as well remove the `ScalarWrapper` in favor of using the previous pattern.

Okay, this might be addressed in a different issue right ? Defining a mutable struct has zero cost ?

**Review by mforets — APPROVED** [review]

**Pull request merged**